src/dict.js

```diff
@@ -69,68 +74,64 @@ Sk.builtin.dict.tp$call = function(args, kw) {
 Sk.abstr.setUpInheritance("dict", Sk.builtin.dict, Sk.builtin.object);
 Sk.abstr.markUnhashable(Sk.builtin.dict);
 
-var kf = Sk.builtin.hash;
-
-Sk.builtin.dict.prototype.key$lookup = function (bucket, key) {
-    var item;
-    var eq;
-    var i;
-
-    // Fast path: We spend a *lot* of time looking up strings
-    // in dictionaries. (Every attribute access, for starters.)
-    if (key.ob$type === Sk.builtin.str) {
-        for (i = 0; i < bucket.items.length; i++) {
-            item = bucket.items[i];
-            if (item.lhs.ob$type === Sk.builtin.str && item.lhs.v === key.v) {
-                return item;
-            }
-        }
-        return null;
+function kf(key) {
+    let key_hash;
+    if (key.$savedKeyHash_) {
+        return key.$savedKeyHash_;
     }
-
-    for (i = 0; i < bucket.items.length; i++) {
-        item = bucket.items[i];
-        eq = Sk.misceval.richCompareBool(item.lhs, key, "Eq");
-        if (eq) {
-            return item;
-        }
+    if (Sk.builtin.checkString(key)) {
+        key_hash = "_" + key.$jsstr();
+        key.$savedKeyHash_ = key_hash;
+        return key_hash;
+    } else if (typeof key === "string") {
+        return "_" + key;
     }
-
-    return null;
+    key_hash = "#_" + Sk.builtin.hash(key).v;
+    key.$savedKeyHash_ = key_hash; // this is a base key hash
```

Comment:
Is it OK to attach this property to _any_ possible key?

Reply:
yes since we already know it's `hashable` since `Sk.builtin.hash` deals with throwing the `unhashable` error here... 

Similarly the `Sk.builtin.hash` algorithm does the same thing...
```javascript
            value.$savedHash_ = value.tp$hash();
            return value.$savedHash_;
```